### PR TITLE
Show #60

### PR DESCRIPTION
**Pull request by allevo**
Add test for showing the #60

### CODE REVIEW

**Comment by mcollina** [comment]
is there any update on this? Would you have time to contribute a fix as well?

**Comment by allevo** [comment]
I'm working on this, but it is not simple. yesterday evening I may find the problem (the code forgets the previous prefix) but I need time to implement it. After that I'll update the PR asap.

If anyone has any suggestions, please come forward!

**Comment by allevo** [comment]
@mcollina @delvedor fixed. Could you review this?

File: index.js [hunk 1]
```diff
@@ -287,7 +291,9 @@ Router.prototype.find = function find (method, path) {
             paramsObj[paramNames[i]] = params[i]
           }
         }
-
+        if (isStatic && ((previousPrefix + previousPath) !== originalPath)) {
```
Comment (delvedor) [inline index.js]: Can you use `isStatic === true`?

File: index.js [hunk 2]
```diff
@@ -317,6 +323,8 @@ Router.prototype.find = function find (method, path) {
     }
     var kind = node.kind
 
+    previousPrefix = previousPrefix + currentNode.prefix
```
Comment (delvedor) [inline index.js]: I think here you can do `previousPrefix += prefix`

**Comment by nwoltman** [comment]
Yes looks like this fixes the problem 👍

**Comment by allevo** [comment]
```
$ git checkout master 
Switched to branch 'master'
$ npm run bench

> find-my-way@1.10.1 bench /Users/allevo/repos/find-my-way
> node bench.js

lookup static route x 16,146,561 ops/sec ±1.94% (81 runs sampled)
lookup dynamic route x 1,102,008 ops/sec ±1.75% (85 runs sampled)
lookup dynamic multi-parametric route x 605,942 ops/sec ±2.16% (85 runs sampled)
lookup dynamic multi-parametric route with regex x 605,571 ops/sec ±1.16% (89 runs sampled)
lookup long static route x 2,588,693 ops/sec ±1.52% (89 runs sampled)
find static route x 17,640,058 ops/sec ±2.59% (82 runs sampled)
find dynamic route x 1,162,933 ops/sec ±3.05% (86 runs sampled)
find dynamic multi-parametric route x 642,138 ops/sec ±2.15% (87 runs sampled)
find dynamic multi-parametric route with regex x 635,571 ops/sec ±1.21% (90 runs sampled)
find long static route x 3,183,752 ops/sec ±1.75% (85 runs sampled)
$ git checkout show-60 
Switched to branch 'show-60'
$ npm run bench

> find-my-way@1.10.1 bench /Users/allevo/repos/find-my-way
> node bench.js

lookup static route x 15,783,816 ops/sec ±1.78% (80 runs sampled)
lookup dynamic route x 1,055,399 ops/sec ±1.67% (85 runs sampled)
lookup dynamic multi-parametric route x 568,765 ops/sec ±2.07% (84 runs sampled)
lookup dynamic multi-parametric route with regex x 532,505 ops/sec ±3.51% (83 runs sampled)
lookup long static route x 1,535,436 ops/sec ±2.79% (80 runs sampled)
find static route x 15,120,369 ops/sec ±1.65% (80 runs sampled)
find dynamic route x 1,083,889 ops/sec ±1.91% (82 runs sampled)
find dynamic multi-parametric route x 580,661 ops/sec ±1.95% (82 runs sampled)
find dynamic multi-parametric route with regex x 541,456 ops/sec ±1.65% (82 runs sampled)
find long static route x 1,879,066 ops/sec ±1.68% (84 runs sampled)
```

Not so good.

IMHO I can avoid to concatenate the prefix storing the absolutePrefix inside the node

File: index.js [hunk 3]
```diff
@@ -267,6 +269,8 @@ Router.prototype.find = function find (method, path) {
   var pindex = 0
   var params = []
   var i = 0
+  var previousPrefix = ''
+  var isStatic = true
```
Comment (mcollina) [inline index.js]: I think `isStatic`  should be reset to true at some point in the logic. If not we are not fixing this bug in its most complete form, as `/static/:parametric/static/:parametric` should be supported as well

**Comment by allevo** [comment]
> can you please restore null? I fear there might be some form of context escaping that might prevent some optimizations.

the restore changes nothing

File: index.js (hunk 3 above)
Comment (allevo) [reply index.js]: me too but the the following test passes with this implementation:
```js

  t.test('with parameter / 5', t => {
    t.plan(2)
    const findMyWay = FindMyWay()

    findMyWay.on('GET', '/bb/:foo/aa/', noop)
    findMyWay.on('GET', '/bb/:foo/aa/bulk', noop)

    t.equal(findMyWay.find('GET', '/bulk'), null)
    t.equal(findMyWay.find('GET', '/bb/foo/bulk'), null)
  })

```

Comment (nwoltman) [reply index.js]: That's not the same as what @mcollina said. That test only has one `:parametric` part (`:foo`).

Comment (allevo) [reply index.js]: I'm bit lost. The following test passes too:
```js

  t.test('with parameter / 6', t => {
    t.plan(3)
    const findMyWay = FindMyWay()

    findMyWay.on('GET', '/static/:parametric/static/:parametric', noop)
    findMyWay.on('GET', '/static/:parametric/static/:parametric/bulk', noop)

    t.equal(findMyWay.find('GET', '/bulk'), null)
    t.equal(findMyWay.find('GET', '/static/foo/bulk'), null)
    t.notEqual(findMyWay.find('GET', '/static/foo/static/bulk'), null)
  })

```

Let me know which test I should add

Comment (mcollina) [reply index.js]: the `isStatic` flag switches only once and it is not reset. I think it should be.

Comment (allevo) [reply index.js]: Me too. But I don't know when and where. Without a failing test is very difficult for me to reset that flag.

**Comment by mcollina** [comment]
Go ahead and add the absolutePrefix thing

**Comment by mcollina** [comment]
> I suppose the performance deterioration is for the string concatenation

The performance deterioration is due to the addition string comparison check (as far I read the code). Maybe it's something else.

File: index.js (hunk 3 above)
Comment (delvedor) [reply index.js]: What does it mean `isStatic`?
The current node is static? The path is  static?

File: index.js [hunk 4]
```diff
@@ -317,23 +323,27 @@ Router.prototype.find = function find (method, path) {
     }
     var kind = node.kind
 
+    previousPrefix += currentNode.prefix
```
Comment (delvedor) [inline index.js]: Here you can directly use `prefix`, it is declared at the beginning of the loop.

File: index.js (hunk 3 above)
Comment (allevo) [reply index.js]: All the node chain (node, its parent and so on) is static. This bug happens only on static route. I'm not able to reproduce it with the url with parameters

File: index.js [hunk 5]
```diff
@@ -306,9 +312,9 @@ Router.prototype.find = function find (method, path) {
     }
 
     var node = currentNode.find(path[0], method)
-    if (node === null) {
+    if (node === ROUTE_NOT_FOUND) {
       node = currentNode.parametricBrother
-      if (node === null) {
+      if (node === ROUTE_NOT_FOUND) {
```
Comment (nwoltman) [inline index.js]: `null` is actually easier to understand here. `ROUTE_NOT_FOUND` sounds like the entire route was not found so the algorithm should return immediately, but here all it means is that it didn't match one of the node types. I'd really suggest reverting all the `ROUTE_NOT_FOUND` changes. `null` is perfectly easy to understand.

File: index.js [hunk 6]
```diff
@@ -306,9 +312,9 @@ Router.prototype.find = function find (method, path) {
     }
 
     var node = currentNode.find(path[0], method)
```
[the remark this answers is not in the document]
Comment (delvedor) [reply index.js]: I think that part of the problem is the node `label` property. We should investigate also there.

**Comment by delvedor** [comment]
Closed by #65.
Thank you anyway!

**Pull request closed without merging**